peers2/topic_pool.go

```diff
@@ -0,0 +1,238 @@
+package peers2
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/discv5"
+	"github.com/status-im/status-go/discovery"
+)
+
+type peerID = discover.NodeID
+
+type TopicPool interface {
+	Topic() discv5.Topic
+	Start(*PeerPool)
+	Stop()
+	ConfirmAdded(peerID) error
+	ConfirmDropped(peerID) error
+}
+
+type satisfiable interface {
+	Satisfied() bool
+}
+
+func IsTopicSatisfied(topic TopicPool) bool {
+	st, ok := topic.(satisfiable)
+	if !ok {
+		// by default, we assume the topic is satisfied
+		return true
+	}
+
+	return st.Satisfied()
+}
+
+func IsAllTopicsSatisfied(topics []TopicPool) bool {
+	for _, topic := range topics {
+		st, ok := topic.(satisfiable)
+		if !ok {
+			continue
+		}
+
+		if !st.Satisfied() {
+			log.Debug("topic not satisfied", "topic", topic.Topic())
+			return false
+		}
+	}
+
+	return true
+}
+
+func SetDiscoverPeriod(p time.Duration) func(*TopicPoolBase) {
+	return func(t *TopicPoolBase) {
+		periodCh := make(chan time.Duration, 1)
+		periodCh <- p
+		t.period = periodCh
+	}
+}
+
+func SetPeersHandler(h FoundPeersHandler) func(*TopicPoolBase) {
+	return func(t *TopicPoolBase) {
+		t.peersHandler = h
+	}
+}
+
+type TopicPoolBase struct {
```

Comment:
Is it supposed to a "default" implementation of topic pool and `TopicPoolWithLimits` an optional one? I.e. what's the use cases for this? `-Base` doesn't seem to fit well, but I'm trying to understand the usage pattern.

Reply:
I will play with this a bit. `Base` suffix is definitely not a Go-way of naming things. `TopicPoolBase` is a default implementation of `TopicPool`, not necessarily useful in the real world scenarios but it nicely encapsulates some logic that would need to be duplicated otherwise.